Issue 36971 match patterns clarifications (#45021)
mdn/content: files/en-us/mozilla/add-ons/webextensions/match_patterns/index.md

Reviewer: ```suggestion
        All browsers match "http" and "https". <br>
```

Reviewer: ```suggestion
          <li>Chrome supports "http", "https", "file", "ftp", "chrome-extension", "ws", "wss", "data", "filesystem" (<a href="/en-US/docs/Web/API/File_and_Directory_Entries_API#history">old FileSystem API</a>), "uuid-in-package" (used by Web Bundles), and "chrome" (browser internals).</li>
```

Reviewer: In Firefox it depends. This list here is the default behavior for match patterns. Some APIs permit anything that looks valid:

- `targetUrlPatterns` in `menus` / `contextMenus` APIs ([source](https://searchfox.org/firefox-main/rev/ac91bfcce1bf3240e2dce40f47c372e76bc4f26c/browser/components/extensions/parent/ext-menus.js#729-734)). For context see https://bugzilla.mozilla.org/show_bug.cgi?id=1280370
- `tabs.onUpdated` `urls` filter ([source](https://searchfox.org/firefox-main/rev/ac91bfcce1bf3240e2dce40f47c372e76bc4f26c/browser/components/extensions/parent/ext-tabs.js#249-257)) (Firefox desktop only)
- `tabs.query` `url` ([source](https://searchfox.org/firefox-main/rev/ac91bfcce1bf3240e2dce40f47c372e76bc4f26c/toolkit/components/extensions/parent/ext-tabs-base.js#2322-2327))

Reviewer: Can you test the current state of support in Chrome? If I need to verify, let me know.

Reviewer: For data:-URL matching, refer to https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts#match_origin_as_fallback

Note: since you are mentioning content script behavior, note that ws, wss, ftp are never matched in practice either, because there is no document that could have these URLs.

Content scripts in Firefox can only match http, https, file, by match pattern.

ws, wss, https, http are the only schemes matching webRequest in Firefox.

So although "data:" is accepted as a value in a match pattern (no errors), it has zero effect in Firefox. In contrast, if `host_permissions` in manifest.json were to contain an unsupported scheme, a warning would be raised. Or if passed to `permissions.request()`, an error is thrown.

Reviewer: This comment was meant to be attached to the line about data:-URL matching in the `<all_urls>` section at https://github.com/mdn/content/blob/41e9122dab9a910db41ffed7d8bc0abbf1ff473c/files/en-us/mozilla/add-ons/webextensions/match_patterns/index.md?plain=1#L143

Reviewer: Release notes (`<>` dummy link?) are not the ideal place for long term documentation. Let's expand https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/extension/isAllowedFileSchemeAccess with special behaviors about file permissions (I thought that we already documented it somewhere?).

For more context that section could also refer to https://blog.mozilla.org/addons/2026/07/23/firefox-153-webextensions-api-updates/

APIs that use match patterns usually accept a list of match patterns, and will perform the appropriate action if the URL matches any of the patterns. See, for example, the [`content_scripts`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts) key in manifest.json.

## Where match patterns are used

Match patterns are used in several places across the manifest.json keys and JavaScript APIs:

- Manifest keys:
  - [`host_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/host_permissions) and [`optional_host_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_host_permissions), to request access to hosts.

>   - [`permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) (Manifest V2 only) and [`optional_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions), which can include host permissions alongside named API permissions.

Reviewer: Firefox accepts match patterns in `optional_permissions`, but that is not cross-browser. In MV3 it is not supposed to support match patterns, but our implementation came after some extensions might already have started using `optional_permissions` to include match patterns. See https://bugzilla.mozilla.org/show_bug.cgi?id=1897580

Perhaps it's worth mentioning that in the article itself if not already done so.

  - The `matches` and `exclude_matches` fields of [`content_scripts`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts) entries.
  - The `matches` field of [`web_accessible_resources`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/web_accessible_resources#manifest_v3_syntax) entries (Manifest V3 only).
  - The `show_matches` and `hide_matches` fields of [`page_action`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/page_action).

> - JavaScript APIs:

Reviewer: Is this list meant to be exhaustive, or giving the impression of it being exhaustive?

It doesn't include Firefox-only MV2-only `contentScripts.register` (and legacy `userScripts.register`). I'm totally fine with NOT including these since they are Firefox-only and MV2-only, just wondering about the intent and whether the outcome matches your intent.

>   - The `origins` property of the object passed to {{WebExtAPIRef("permissions.request()")}} and returned by other {{WebExtAPIRef("permissions")}} methods.

Reviewer: "the object" is described by https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/Permissions

>   - The `matches` and `excludeMatches` properties passed to {{WebExtAPIRef("scripting.registerContentScripts()")}}.

Reviewer: ... and also returned by various methods in the scripting API. I'm wondering, how do you decide whether or not to include the mention? Above, you noted the other permission API methods, but here you don't.

I have no strong opinions right now, just wondering because of the apparent inconsistency.

>   - The `matches` and `excludeMatches` properties passed to {{WebExtAPIRef("userScripts.register()")}}.

Reviewer: Similarly here, there are multiple methods in userScripts where the concept is relevant: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/userScripts/RegisteredUserScript

APIs that use match patterns usually accept a list of match patterns, and will perform the appropriate action if the URL matches any of the patterns. See, for example, the [`content_scripts`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_scripts) key in manifest.json.

## Where match patterns are used

Match patterns are used in several places across the manifest.json keys and JavaScript APIs:

- Manifest keys:
  - [`host_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/host_permissions) and [`optional_host_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_host_permissions), to request access to hosts.

>   - [`permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) (Manifest V2 only) and [`optional_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions) (Firefox only for Manifest V3), which can include host permissions alongside named API permissions.

Reviewer: Text was suggesting that optional_permissions was MV3-only, that is not the case. Also linking to https://bugzilla.mozilla.org/show_bug.cgi?id=1897580 to emphasize its uncertain future.

```suggestion
  - [`permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) (Manifest V2 only) and [`optional_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions) (Manifest V2 only, also supported [but deprecated](https://bugzilla.mozilla.org/show_bug.cgi?id=1897580) in MV3 in Firefox only), which can include host permissions alongside named API permissions.
```

>   - [`permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) (Manifest V2 only) and [`optional_permissions`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/optional_permissions), which can include host permissions alongside named API permissions.

Reviewer: This line is duplicated?

```suggestion
```

    <tr>
      <td><code>*</code></td>
      <td>

>         All browsers match "http" and "https". <BR>

Reviewer: ```suggestion
        All browsers match "http" and "https".<br>
```

### \<all_urls>

The special value `<all_urls>` matches all URLs under the browser-supported schemas, where:

- **Firefox** supports "http", "https", "ws", "wss", "ftp", "data", and "file". However, while `data:` URLs are matched, they aren't used for content script or stylesheet injection.
- **Chrome** supports "http", "https", "ftp", "file", and "data".

> - **Safari** only supports "http" and "https".

Reviewer: ```suggestion
- **Firefox** matches "http", "https", "ws", "wss", "ftp", "data", and "file". However, while `data:` URLs are matched, they aren't used for content script or stylesheet injection.
- **Chrome** matches "http", "https", "ftp", "file", and "data".
- **Safari** only matches "http" and "https".
```
